networkx/utils/decorators.py

```diff
@@ -454,22 +365,539 @@ def random_array(dims, random_state=1):
     np_random_state
     """
 
-    @decorator
-    def _random_state(func, *args, **kwargs):
-        # Parse the decorator arguments.
-        try:
-            random_state_arg = args[random_state_index]
-        except TypeError as e:
-            raise nx.NetworkXError("random_state_index must be an integer") from e
-        except IndexError as e:
-            raise nx.NetworkXError("random_state_index is incorrect") from e
+    return argmap(create_py_random_state, random_state_index)
+
+
+class argmap:
+    """A decorating class which applies a map to a function's arguments before
+    calling it.
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply to arguments
+
+    *params: iterable of (int, str or tuple)
+        A list of parameters, specified either as strings (their names), ints
+        (numerical indices) or arbitrarily nested tuples thereof.
+
+    Examples
+    --------
+    The decorated function
+
+        @argmap(sum, 'x', 2)
+        def foo(x, y, z):
+            return x, y, z
+
+    is equivalent to
+
+        @argmap(sum, 'z', 0)
+        def foo(x, y, z):
+            return x, y, z
+
+    or
+
+        def foo(x, y, z):
+            x = sum(x)
+            z = sum(z)
+            return x, y, z
+
+    Transforming functions can be applied to multiple arguments, such as
 
-        # Create a numpy.random.RandomState instance
-        random_state = create_py_random_state(random_state_arg)
+        def swap(x, y):
+            return y, x
 
-        # args is a tuple, so we must convert to list before modifying it.
-        new_args = list(args)
-        new_args[random_state_index] = random_state
-        return func(*new_args, **kwargs)
+        @argmap(swap, ('a', 'b')):
+        def foo(a, b, c):
+            return a, b, c
 
-    return _random_state
+    is equivalent to
+
+        def foo(a, b, c):
+            a, b = swap(a, b)
+            return a, b, c
+
+    Also, transforming functions can preceed a try-finally block, if they both
+    transform an argument and return a closing function:
+
+        def open_file(fn):
+            f = open(fn)
+            return f, lambda: f.close()
+
+        @argmap.try_finally(open_file, 'file')
+        def foo(file):
+            print(file.read())
+
+    is equivalent to
+
+        def foo(file):
+            file, close_file = open_file(file)
+            try:
+                print(file.read())
+            finally:
+                close_file()
+
```

Comment:
add a historical note about why this class exists, and see-also the decorators which are built on it